### ISSUE TITLE
Pre-processing chunks and buffers

### ISSUE THREAD

**Issue by ThatRendle** [flowjs/flow.js#42]
First: love _flow.js_, especially _ng-flow_, great job.

Question: I need to pre-process the actual buffers for chunks before they're uploaded, with a couple of possible actions:
- supply a new buffer;
- skip the chunk altogether.

I can't see any way to do this at the moment, but have I missed something, or do I need to fork and if so, would you be interested in a PR with this functionality?

Suggested use cases:
- Compressing data with something like js-deflate before sending;
- Identifying "blank" chunks (e.g. all zero-bytes, like in a disk image) and either skipping or replacing with an empty buffer

**Comment by hellais** [comment]
@markrendle I am trying to achieve basically your same goal, but in my case setting the chunk size to infinity since this would be a very big overhead for large files.

I posted [here](https://github.com/flowjs/flow.js/issues/9#issuecomment-74690158) some possible ways of achieving this goal.

@AidasK thanks for this bit of code, it is quite useful to see how the various pieces of flow.js fit together.

**Comment by hellais** [comment]
@AidasK @markrendle this is what I had in mind to add to flow.js: https://github.com/hellais/flow.js/commit/1941170b2e91efd58a247a3b4f9ef951b9657ae6

Would you accept a patch that implements something along those lines?

**Comment by AidasK** [comment]
Yea, looks great, do all the tests pass? I would like to avoid BC.

Would it be wise to make a flowjs built in implementation of `fileStreamFactory`? By doing so, we could eliminate all `if (this.stream) {`, `if (typeof this.flowObj.fileStreamFactory === "function") {` checks and code should look a bit cleaner.

**Comment by hellais** [comment]
@AidasK yes I was thinking of doing that, but wanted to hear your feedback first.

I could make a simple wrapper around the FileAPI implemented as a stream.Readable.

There will still be a little bit of code like that because in my use case the stream is not seekable, but I think the overall code quality would increase.

The other drawback is that you would have to add as a dependency the browserify shim for [readable-stream](https://github.com/iojs/readable-stream)

**Comment by AidasK** [comment]
`The other drawback is that you would have to add as a dependency the browserify shim for readable-stream.` Would it be heavy? Or just a few lines of code? Can you make a small code snippet of it?

**Comment by hellais** [comment]
@AidasK it's not a large dependency and it would mean adding it to the dependencies inside of npm and instructing the grunt build script to bake it into the build of flow.js.

You can find the code for it here: https://github.com/iojs/readable-stream.

It's pretty well written and pretty widely used.

**Comment by AidasK** [comment]
Do you mean we should add a whole `readable-stream` repo to the flow.js build? It takes more than a few lines of code and I think it is too big.

Lets keep it simple. How about we create a file factory? It would be used on top of the file.
https://github.com/flowjs/flow.js/blob/master/src/flow.js#L695
```js
 this.file = flowObj.opts.fileFactory(file);
```
Default implementation:
```js
flowObj.fileFactory = function (file) {
return {
name: file.fileName || file.name,//firefox compatibility
size: file.size,
relativePath: file.relativePath || file.webkitRelativePath || file.fileName || file.name,
slice: file[(file.slice ? 'slice' :
        (file.mozSlice ? 'mozSlice' :
          (file.webkitSlice ? 'webkitSlice' :
            'slice')))].bind(file)
};
} 
```

Customized implementation for your case:
```js
flowObj.fileFactory = function (file) {
var stream = streamFactory(file);//your factory
return {
name: file.fileName || file.name,
size: stream.size,
relativePath: file.relativePath || file.webkitRelativePath || file.fileName || file.name,
slice: function (startByte, endByte, type) {return stream.read(endByte - startByte)}
};
} 
```

What do you think? In this way logic would be simplified and we would not be tied to stream interface.

**Comment by hellais** [comment]
@AidasK yes I like this idea! It's adds an extra layer of indirection, but I guess it is justified by the fact that this way the dependency on streams is optional.

I will work on (probably next week) integrating this suggestion into the above proposed code.

Thanks a lot for your responsiveness and flexibility in adding new features to your library. Unfortunately it's not something as common as you would think in free software projects and it warms my :heart: :)

**Comment by AidasK** [comment]
Glad you like it, this library definitely needs some warmth

**Comment by AidasK** [comment]
Third solution. How about just moving a slice method to the config?
https://github.com/flowjs/flow.js/blob/master/src/flow.js#L1290
```js
evalOpts(this.flowObj.opts.slice, startByte, endByte, type, this.fileObj, this);
```
And in config:
```js 
this.defaults: {
....
  slice: function (startByte, endByte, type) {
    this.file[(this.file.slice ? 'slice' :
        (this.file.mozSlice ? 'mozSlice' :
          (this.file.webkitSlice ? 'webkitSlice' :
            'slice')))](startByte, endByte, type)
  }
  ....
}
```
What do you think? This is just a slight change, but should work for your case.

**Comment by hellais** [comment]
@AidasK so while implementing this I realised that there is a problem with this approach. Node streams work asynchronously, meaning that when you call the `slice` method you will not get back data immediately, but have to wait for a certain event to fire and only then will you be able to write the data.

For this reason I think the slice method should have an extra argument called something like callback that should be fired with the content of the sliced data.

Do you think this is a good idea? Any better ways of doing this? Drawbacks?

Yes I am working on a separate branch (feature/fileTransform) and once it's ready to be merged I will rebase it with master to make code review easier.

**Comment by AidasK** [comment]
Maybe we should make a slice of a file in a preprocess state? It already supports async callbacks. We only need to move our slice method to it. However it might cause a BC for some users, who are using this feature. Any ideas how BC can be avoided? (In worst case 3.0.0)

Too bad we can not use pormises http://caniuse.com/#feat=promises

Any suggestions?

**Comment by hellais** [comment]
Hum, yes the idea of using preprocess sounds viable, I will look into how it can be done.

And yes I agree that it's a pitty we can't use promises, I quite like that pattern. There are some polyfills that could be used for older browsers, but that requires adding build dependencies to it.

BTW, are you IRC or jabber? We could discuss this more there: I am on OFTC as hellais and on jabber as hellais@jabber.ccc.de

**Comment by hellais** [comment]
We could add another state in the preprocess state machine (i.e. making the states go from 0-3 instead of 0-2).

There would then be another, overridable, method called `read` that takes care of reading the data from the file object. If it's implemented in the `preprocessFinished` callback we should wait to call `this.send()` and instead call `this.read`.
If it's not implemented it should advance the state machine to 3 and call `this.send()`.

The implementer of `read` should call `readFinished` when it has done reading the correct number of bytes.
`readFinished` will then work like `preprocessFinished` and advance the state machine to 3 and call `this.send()`.

What do you think of this approach?

**Comment by AidasK** [comment]
Sounds good to me. I am leaving promises for the next major version of flow.js, but it is a must. I have connected OFTC, if you have any questions, please ask. 
Thanks for your time, you are doing a great progress!

**Comment by evilaliv3** [comment]
@AidasK: i've analized a little bit the patch by @hellais (https://github.com/flowjs/flow.js/pull/82) and ii would like to discuss with you a little bit in order to find out with you a possible fast solution for this ticket and it's integration as it is really blocking us on the GlobaLeaks project.

basically @hellais already described what we would like to do:
given flow handling of the real file, we would like to encrypt the payload send it using flow so that we where thinking to use the preprocess functionality.
given that we are applying encryption (but other probably would like to apply compression or whatever) we have to keep in mind that the size of the payload to be sent is different from the original sizeof the file (e.g. in our case it's bigger) and that given how flowjs protocol works the size should be know in advance (that it's real in our case but real in general for a compressor).

so probably we should design the possiblity for flow to read the file throught an abstraction of the file similar to how @hellais have done but that would decuble more the actions of the preoprocessing from the read need of flowjs eventually with the usage of promises (instead noew in the patch of @hellais each call to read() corresponds to a preoprocessing action not so well decoupled)

**Comment by AidasK** [comment]
I have left some comments on a pull request. Basically all what we need is some documentation about new features and example how we could use it with compression. I'll try to be more responsive to merge this feature in.

**Comment by evilaliv3** [comment]
thanks @AidasK 

for sure we will provide an example and the documentation but i would like to discuss with you in order to finalize the solution. are you available for a short call on skype? (my nick is evilaliv3)

**Comment by evilaliv3** [comment]
by the way the patch by @hellais is quite complete but i think it would require some refinement in agreement with the flowjs directions and i've some suggestions for flowjs itself.

the issue now is that flowjs operates using the length he knows that is the one of the original file and is able to perform callback to user defined function using this size; in addition it expect the user defined functions to send out a payload big as the chunk size and a precise total size know at the start. this last precondition is ok only for some encryption methods that knows the size of the encrypted message in advance but  is not for sure usable by compressors algorithms that will know it only at the end of the compression.

so, what i would propose is to edit a little bit the flowjs algorithm so that it wont' require this size to be known in advance and simply relay on an ending marker for saying this is the last chunk (a thing similiar let say to the chunk encoding of http). what do you think?

**Comment by AidasK** [comment]
Lets add this feature to a pull request. My only request is to avoid breaking changes if possible.

**Comment by evilaliv3** [comment]
sure.
ok so lets bypass for now what i've proposed and implement this simple thing:
1. i would implement a getSize() callback overridable by the user so that flojws will always work knowing the endign size
2. given 1. the read callback by hellais will make the user function fetch a chunk but the start/endbyte sizes will be already refer to the ending size.
3. nothing else to go :)

what do you think?

**Comment by AidasK** [comment]
Looks ok, lets try.

**Comment by evilaliv3** [comment]
the points discussed (mainly the fact that flowjs need to work knowing the final size) is due to the fact that flowjs should know the number of chunks at start in order to be able to provide time estimations.

this for a lot i think would remain a limit of the flowjs library (already present and not added by our patch).
